wasmcloud_host/lib/wasmcloud_host/lattice/state_monitor.ex

```diff
@@ -195,22 +195,61 @@ defmodule WasmcloudHost.Lattice.StateMonitor do
     %State{state | providers: providers}
   end
 
-  # This map is keyed by provider public key, which then contains a sub-map
-  # keyed by link name. The value corresponding to the link name key is a list
-  # of hosts on which that provider-link combo is running.
+  # This map is keyed by provider public key, which contains a list of
+  # maps with the keys "link_name", "contract_id", and "host_ids", as
+  # shown below.
   #
   # %{
-  #    "Vxxxxxx": %{
-  #       "default" => ["Nxxxx"],
-  #       "special" => ["Nxxxx"]
-  #    }
+  #     "Vxxxx":  [
+  #        %{
+  #            "link_name": "default",
+  #            "contract_id": "wasmcloud:keyvalue",
+  #            "host_ids": ["Nxxxx"]
+  #         },
+  #        %{
+  #            "link_name": "special",
+  #            "contract_id": "wasmcloud:keyvalue",
+  #            "host_ids": ["Nxxxx"]
+  #         },
+  #    ]
   # }
   def add_provider(pk, link_name, contract_id, host, previous_map) do
-    provider_map = Map.get(previous_map, pk, %{})
-    link_map = Map.get(provider_map, link_name, %{})
-    hosts_list = Map.get(link_map, link_name, [])
-    link_map = Map.put(link_map, link_name, [host | hosts_list])
-    Map.put(provider_map, pk, link_map)
+    # This logic will add in a new entry for every call
+    # It shouldn't add information if it's a duplicate, and it should only
+    # append a host_id to the list if the link_name + contract_id already exists
+    previous_instances = Map.get(previous_map, pk, [])
+
+    # Check for existence of link_name and contract_id pair
+    # [%{}] results in single el
```

Comment:
what is a  "single el" ?

Reply:
Whoops, those are comments that I left for myself. I'll clean that up